lib/ai/model/tool.go

```diff
@@ -109,18 +134,281 @@ func (*commandExecutionTool) parseInput(input string) (*commandExecutionToolInpu
+				return trace.Wrap(ctx.Err())
+			}
+		}
+
+		return nil
+	}
+	searchAndAppendPlain := func(resourceType string) error {
+		return searchAndAppend(resourceType, func(resource types.Resource) (promptResource, error) {
+			return promptResource{
+				Name:    resource.GetName(),
+				Kind:    resource.GetKind(),
+				SubKind: resource.GetSubKind(),
+				Labels:  resource.GetMetadata().Labels,
+			}, nil
+		})
+	}
+
+	go searchAndAppend(types.KindNode, func(resource types.Resource) (promptResource, error) {
+		return promptResource{
+			Name:         resource.GetName(),
+			Kind:         resource.GetKind(),
+			SubKind:      resource.GetSubKind(),
+			Labels:       resource.GetMetadata().Labels,
+			FriendlyName: resource.(types.Server).GetHostname(),
+		}, nil
+	})
+	go searchAndAppendPlain(types.KindApp)
+	go searchAndAppendPlain(types.KindKubernetesCluster)
+	go searchAndAppendPlain(types.KindDatabase)
+	go searchAndAppendPlain(types.KindWindowsDesktop)
+
+	sb := strings.Builder{}
+	total := 0
+	for resource := range foundResources {
+		yaml, err := yaml.Marshal(resource)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+
+		sb.WriteString(string(yaml))
+		sb.WriteString("\n")
+
+		total++
+		if total >= maxShownRequestableItems {
+			break
+		}
+	}
+
+	if sb.Len() == 0 {
+		return "No requestable resources found", nil
+	}
+
+	return sb.String(), nil
+}
+
+type promptResource struct {
+	Name         string            `yaml:"name"`
+	Kind         string            `yaml:"kind"`
+	SubKind      string            `yaml:"subkind"`
+	Labels       map[string]string `yaml:"labels"`
+	FriendlyName string            `yaml:"friendly_name,omitempty"`
+}
+
+type accessRequestCreateTool struct{}
+
+func (*accessRequestCreateTool) Name() string {
+	return "Create Access Requests"
+}
+
+func (*accessRequestCreateTool) Description() string {
+	return fmt.Sprintf(`Create an access request with a set of roles to, a set of resource UUIDs, a reason, and a set of suggested reviewers.
+A valid access request must be either for one or more roles or for one or more resource IDs.
+If the user is not specific enough, you may try to determine the correct roles or resource UUIDs by any means you see fit.
+
+The input must be a JSON object with the following schema:
+
+%vjson
+{
+	"roles": []string, \\ The optional set of roles being requested
+	"resources": []{
+		"type": string, \\ The resource type
+		"id": string, \\ The resource name
+		"friendlyName": string \\ Optional display-friendly name for the resource
+	}, \\ The optional set of UUIDs for resources being requested
+	"reason": string, \\ A reason for the request. This cannot be made up or inferred, it must be explicitly said by the user
+	"suggested_reviewers": []string \\ An optional list of suggested reviewers; these must be Teleport usernames
+}
+%v
+`, "```", "```")
+}
+
+func (*accessRequestCreateTool) Run(ctx context.Context, toolCtx *ToolContext, input string) (string, error) {
```

Comment:
I'm really questioning the purpose of this method in the interface. Most of the implementations are "not implemented"

Reply:
From what I understand we have two different tools:
- actionTools that do something and end the agent loop
- lookupTools that are here to feed more information to the agent

The current tool implementation only support lookup tools, and we end up with a bunch of unimplemented Run() methods and special cases are leaking into the main agent loop.

When we'll clean this part of the code it might be useful to have a proper ActionTool type that ends the execution so can support command execution, access request creation, and sql query generation without awkward empty functions.